### PR TITLE
AL/MIN-503. Liquidation impl functions.

### PR DESCRIPTION
**Pull request by lut1k**
**Description:**

This PR implements some functions for liquidation. The liquidation architecture is also built in this PR.
In this PR, it is necessary to check the architectural solutions of the liquidation flow. Also check the implemented functions.

Implemented functions for liquidation flow and covered with unit-tests:

Controller:
- `fn get_all_users_with_insolvent_loan()` - calculates and gets all insolvent loans of users in the protocol. Calls a function internally `accrue_interest_rate`. To determine that the loan is insolvent calls the function `get_hypothetical_account_liquidity`, if the shortfall is greater than zero, then such loan is insolvent. Returns: returns a unique collection of users with insolvent loan (as a btree set).

Excluded seprate module `liquidation`. This module provides functionality for handling user loans. Implemented methods:
- `fn build_user_loan_state()` -  Constructs the user's state of loan based on the current state of the storage on the blockchain. This function internally calls functions from the controller pallet,  which internally call `accrue_interest_rate`.
- calculators for total amounts;
- `fn choose_liquidation_mode()` - selects the liquidation mode for the user's loan. The choice of the liquidation mode is made based on the parameters of the current number of user's liquidation attempts and the current state of the user's loan.

Functions that have a raw implementation and are not covered by unit-tests. These functions don't need to be review:
- `fn process_insolvent_loan()`;
- `fn try_mutate_attempts()`;

**Changes in storage or API:**

--

**Checklist**

- [x] Code is covered with tests
- [x] Code is covered with benchmarks
- [x] Fresh benchmark results included

### CODE REVIEW

File: pallets/controller/src/lib.rs [hunk 1]
```diff
@@ -666,7 +666,7 @@ impl<T: Config> ControllerManager<T::AccountId> for Pallet<T> {
 	///          hypothetical account shortfall below collateral requirements).
 	fn get_hypothetical_account_liquidity(
 		account: &T::AccountId,
-		underlying_to_borrow: CurrencyId,
+		underlying_to_borrow: Option<CurrencyId>,
```
Comment (dev-vadim) [inline pallets/controller/src/lib.rs:669]: Have you considered creating the separate function for the case when we don`t need underlying_to_borrow?

File: pallets/risk-manager/src/lib.rs [hunk 2]
```diff
@@ -385,19 +345,24 @@ impl<T: Config> Pallet<T> {
...
-				LiquidationMode::Complete => Self::calculate_complete_liquidation(&borrower, &user_loan_state)
-					.map_err(|_| OffchainErr::CheckFail)?,
-				LiquidationMode::ForgivableComplete => {
-					Self::calculate_forgivable_complete_liquidation(&borrower, &user_loan_state)
-						.map_err(|_| OffchainErr::CheckFail)?
-				}
-			};
+		let user_loan_state: UserLoanState<T> =
+			UserLoanState::build_user_loan_state(&borrower).map_err(|_| OffchainErr::CheckFail)?;
+		ensure!(borrower == *user_loan_state.get_user(), OffchainErr::CheckFail);
```
Comment (dev-vadim) [inline pallets/risk-manager/src/lib.rs:350]: What does CheckFail mean? Maybe rename it to something more informative?

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -881,7 +883,18 @@ impl<T: Config> ControllerManager<T::AccountId> for Pallet<T> {
 		Self::controller_params(pool_id).protocol_interest_threshold
 	}
 
-	/// TODO: Raw implementation. Cover with unit-tests.
+	/// Calculates the amount of collateral based on the parameters pool_id and the supply amount.
+	/// Reads the collateral factor value from storage.
+	///
+	/// Cannot overflow, because the collateral factor is always less than one.
+	/// Returns: `collateral_amount = supply_amount * collateral_factor`.
+	fn calculate_collateral(pool_id: CurrencyId, supply_amount: Balance) -> Balance {
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:893]: This is not a pool collateral. This is a function for mathematical calculations of any balance based on the collateral_factor of a particular liquidity pool.

File: pallets/controller/src/lib.rs (hunk 1 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs:671]: I did not consider this option :)

File: pallets/risk-manager/src/lib.rs (hunk 2 above)
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:351]: I have not corrected it now. Added TODO

Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:353]: I added this check to make sure we are calling the liquidation of the correct user.

**Review by Apollo50 — APPROVED** [review]

**Review by dev-vadim — APPROVED** [review]

**Pull request merged**